### PR TITLE
A few fixes for nsenter

### PR DESCRIPTION
**Pull request by avagin**

### CODE REVIEW

File: namespaces/nsenter/nsenter.c [hunk 1]
```diff
@@ -26,6 +26,11 @@ void get_args(int *argc, char ***argv)
 {
 	// Read argv
 	int fd = open("/proc/self/cmdline", O_RDONLY);
+	if (fd < 0) {
+		fprintf(stderr,
+			"nsenter: Unable to open /proc/self/cmdline: %m\n");
```
Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: Does %m also work with other C compilers like clang for e.g.?

File: namespaces/nsenter/nsenter.c [hunk 2]
```diff
@@ -156,22 +166,34 @@ void nsenter()
...
-		int fd = open(buf, O_RDONLY);
-		if (fd == -1) {
-			// Ignore nonexistent namespaces.
+		// A zombie process has links on namespaces, but they can't be opened
+		struct stat st;
+		if (fstatat(ns_dir_fd, namespaces[i], &st, AT_SYMLINK_NOFOLLOW) == -1) {
 			if (errno == ENOENT)
 				continue;
+			fprintf(stderr,
+				"nsenter: Failed to stat ns file \"%s\" for ns \"%s\" with error: \"%s\"\n",
```
Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: Might want to use %m here also for consistency.

File: namespaces/nsenter/nsenter.c (hunk 1 above)
Comment (avagin) [inline namespaces/nsenter/nsenter.c]: [avagin@localhost ~]$ cat test.c 
#include <stdio.h>

int main()
{
	printf("%m\n");
	return 0;
}
[avagin@localhost ~]$ clang -Wall test.c 
[avagin@localhost ~]$ ./a.out 
Success
[avagin@localhost ~]$ gcc -Wall test.c 
[avagin@localhost ~]$ ./a.out 
Success

Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: cool, thanks!

File: namespaces/nsenter/nsenter.c [hunk 3]
```diff
@@ -156,22 +166,34 @@ void nsenter()
...
+				ns_dir, namespaces[i], strerror(errno));
+			exit(1);
+		}
 
+		int fd = openat(ns_dir_fd, namespaces[i], O_RDONLY);
+		if (fd == -1) {
 			fprintf(stderr,
 				"nsenter: Failed to open ns file \"%s\" for ns \"%s\" with error: \"%s\"\n",
```
Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: %m here as well?

Comment (mrunalp) [inline namespaces/nsenter/nsenter.c]: Use %m here as well?

Comment (avagin) [inline namespaces/nsenter/nsenter.c]: I have uploaded a few patches to use %m everywhere. thanks.

File: namespaces/nsenter/nsenter.c [hunk 4]
```diff
@@ -156,51 +162,60 @@ void nsenter()
...
 			exit(1);
 		}
 		close(fd);
 	}
+	close(ns_dir_fd);
 
 	// We must fork to actually enter the PID namespace.
 	int child = fork();
+	if (child == -1) {
+		fprintf(stderr, "nsenter: Unable to fork a process: %m\n");
```
Comment (avagin) [inline namespaces/nsenter/nsenter.c]: Fixed. Thanks.

**Comment by vmarmol** [comment]
LGTM

**Comment by mrunalp** [comment]
Needs rebase.

**Comment by mrunalp** [comment]
@avagin Sorry for bothering you again but it now includes a merged commit. :/ Could you fix that and squash some  of your commits?

**Comment by avagin** [comment]
@mrunalp Sorry, it was my fault. I have updated.

File: namespaces/nsenter/nsenter_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	args := []string{"nsenter-exec", "--nspid", fmt.Sprintf("%d", os.Getpid())}
+
+	cmd := &exec.Cmd{
+		Path: os.Args[0],
+		Args: args,
+	}
+
+	err := cmd.Run()
+	if err != nil {
+		t.Fatal("nsenter exits with a zero exit status")
```
Comment (mrunalp) [inline namespaces/nsenter/nsenter_test.go]: Should be non-zero exit status here in the message?

**Comment by mrunalp** [comment]
drone build is missing for this one. @crosbymichael Any ideas?

File: namespaces/nsenter/nsenter_test.go (hunk 5 above)
Comment (avagin) [inline namespaces/nsenter/nsenter_test.go]: Fixed. Thanks.

**Comment by mrunalp** [comment]
Drone came back! Also tested locally. LGTM.

**Pull request merged**